pftools/python/parflow/tools/database/core.py

```diff
@@ -651,6 +648,52 @@ def _process_dynamic(self):
                 if name is not None:
                     self.__dict__[name] = klass(self)
 
+    # ---------------------------------------------------------------------------
+
+    def _get_value_from_location(self, location='.'):
+        """
+        Internal function to get the value, container and key from location
+        """
+        container, key = self._get_container_and_key(location)
+
+        if key is None:
+            value = container
+        else:
+            value = getattr(container, key, None)
+
+        if isinstance(value, PFDBObj):
+            value = getattr(value, '_value_', None)
+
+        return value, container, key
+
+    # ---------------------------------------------------------------------------
+
+    def _get_container_and_key(self, location):
+        """
+        Internal function to get the container and key from location
+        """
+        # Normalize the location so we don't have to worry about
+        # complicated paths like "a/b/../c/./."
+        location = os.path.normpath(location)
+
+        split = location.split('/')
+        path, key = '/'.join(split[:-1]), split[-1]
+
+        if key == '.':
+            return self, None
+
+        parent = self
+        while key == '..' and getattr(parent, '_parent_', None):
```

Comment:
hum, it might be right, but does the key can be `..`. I thought those `..` will be at the head of the path...

Reply:
So, because of the way I'm splitting the path, whatever is after the last `/` is considered the key. But it is not really a key if it is `..`.

This while loop is mostly to handle things like if the user put in `../..`. If they do that, the key will be `..`, which is not really a key.

For something like `Geom/Porosity/..`, that automatically gets turned into `Geom` from the `os.path.normpath` call.

For something like `../Porosity`, that `..` will get handled by `get_selection_from_location()` call.